ble_pricetags/services.py

```diff
@@ -0,0 +1,288 @@
+    if image.size != (canvas_width, canvas_height):
+        _LOGGER.debug(f"Original image size: {image.size}. Scaling to ({canvas_width}, {canvas_height})...")
+        image = image.resize((canvas_width, canvas_height))
+        _LOGGER.debug("Scaling complete.")
+    else:
+        _LOGGER.debug("Image already matches target dimensions.")
+    return image
+
+def notification_handler(sender, data):
+    global image_part_size, upload_done
+    hex_data = data.hex()
+    _LOGGER.debug(f"Got bytes: {hex_data}")
+    if hex_data.startswith("01"):
+        size = int.from_bytes(data[1:3], byteorder="little")
+        image_part_size = size - 4
+        _LOGGER.debug(f"Display requested part size: {image_part_size}")
+    elif hex_data.startswith("02"):
+        asyncio.run_coroutine_threadsafe(send_command("03"), asyncio.get_event_loop())
+    elif hex_data.startswith("05"):
+        if data[1] == 0x08:
+            _LOGGER.debug("Upload complete. Disconnecting.")
+            upload_done = True
+            asyncio.run_coroutine_threadsafe(ble_client.disconnect(), asyncio.get_event_loop())
+        elif data[1] != 0x00:
+            _LOGGER.debug("Display reported an error during upload.")
+        else:
+            ack_part = struct.unpack("<I", data[2:6])[0]
+            asyncio.run_coroutine_threadsafe(send_next_image_part(ack_part), asyncio.get_event_loop())
+
+async def send_next_image_part(part_number):
+    global img_data
+    start = part_number * image_part_size
+    end = start + image_part_size
+    chunk = img_data[start:end]
+    if not chunk:
+        return
+    header = struct.pack("<I", part_number)
+    full_packet = header + chunk
+    try:
+        await ble_client.write_gatt_char(CHARACTERISTIC_IMG_UUID, full_packet)
+    except Exception as e:
+        _LOGGER.warning(f"Error sending part {part_number}: {e}")
+
+async def send_command(cmd_hex):
+    if ble_client and ble_client.is_connected:
+        data = bytes.fromhex(cmd_hex)
+        await ble_client.write_gatt_char(CHARACTERISTIC_CMD_UUID, data)
+
+async def scan_and_find_device(target_mac_addr, timeout=10.0):
+    """
+    Perform a Bluetooth scan and find a device with the specified MAC address.
+    
+    Args:
+        target_mac_addr (str): The MAC address of the target device.
+        timeout (float): The timeout for the scan in seconds.
+
+    Returns:
+        target_device (BleakScanner.Device): The found device object, or None if not found.
+    """
+    try:
+        devices = await asyncio.wait_for(BleakScanner.discover(timeout=timeout), timeout=timeout + 10.0)
+        _LOGGER.debug(f"Scan complete, found {len(devices)} devices.")
+    except asyncio.TimeoutError:
+        _LOGGER.error("BLE scan timed out.")
+        return None
+
+    if not devices:
+        _LOGGER.debug("No devices found.")
+        return None
+
+    _LOGGER.debug(f"Found {len(devices)} BLE devices:")
+    for device, advertisement_data in devices:
```

Comment:
I get an error that devices isn't a tuple and the code fails.
```suggestion
    for device in devices:
```